### PR TITLE
Adopt safe model and trial saving practices in the multi-worker setting

### PR DESCRIPTION
**Pull request by jamesmullenbach**
Based on the [reference guide](https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras#model_saving_and_loading) for saving models when training with multiple workers.

### CODE REVIEW

File: keras_tuner/engine/oracle.py [hunk 1]
```diff
@@ -233,6 +233,8 @@ def update_trial(self, trial_id, metrics, step=0):
                 )
                 trial.metrics.register(metric_name, direction=direction)
             trial.metrics.update(metric_name, metric_value, step=step)
+        if self.should_report:
+            self._save_trial(trial)
         self._save_trial(trial)
```
Comment (haifeng-jin) [inline keras_tuner/engine/oracle.py:238]: Should this line be removed?

**Review by haifeng-jin — COMMENTED** [review]
Thanks for the PR!

File: keras_tuner/engine/tuner_utils.py [hunk 2]
```diff
@@ -201,6 +206,10 @@ def on_epoch_end(self, epoch, logs=None):
         if self.objective.better_than(current_value, self.best_value):
             self.best_value = current_value
             self.model.save_weights(self.filepath)
+            # Remove temporary saved model files from non-chief workers.
+            ds_utils.remove_temp_dir_with_filepath(
+                self.filepath, self.model.distribute_strategy
+            )
```
Comment (haifeng-jin) [inline keras_tuner/engine/tuner_utils.py:212]: Two questions about this part:
1. Why do we only remove the temp dir when we got a better result? Do we do it only when saving weights? However we also did a save_weights on line 203.
2. In the doc of `remove_temp_dir_with_filepath()`, the first arg should be the original file path, but it seems to be using the temporary file path assigned on line 197.

**Review by haifeng-jin — COMMENTED** [review]
Please also check the linting errors with `sh shell/lint.sh`.
Thanks.

File: keras_tuner/engine/oracle.py (hunk 1 above)
Comment (jamesmullenbach) [reply keras_tuner/engine/oracle.py]: Yes, thanks!

**Review by haifeng-jin — APPROVED** [review]
LGTM

**Pull request merged**